src/storage/txn/commands/check_txn_status.rs

```diff
@@ -874,10 +959,113 @@ pub mod tests {
+        let v = b"v1";
+        let ts = TimeStamp::compose;
+
+        // Check with resolving_pessimistic_lock flag.
+        // Path: there is no commit or rollback record, no rollback record should be written.
+        must_success(
+            &engine,
+            k,
+            ts(3, 0),
+            ts(3, 0),
+            ts(4, 0),
+            true,
+            false,
+            true,
+            |s| s == LockNotExistDoNothing,
+        );
+        must_get_none(&engine, k, ts(5, 0));
+
+        // Path: there is no commit or rollback record, error should be reported if
+        // rollback_if_not_exist is set to false.
+        must_err(&engine, k, ts(3, 0), ts(5, 0), ts(5, 0), false, false, true);
+
+        // Path: the pessimistic primary key lock dose exist, and it's not expired yet.
+        must_acquire_pessimistic_lock_with_ttl(&engine, k, k, ts(10, 0), ts(10, 0), 10);
+        must_pessimistic_locked(&engine, k, ts(10, 0), ts(10, 0));
+        must_success(
+            &engine,
+            k,
+            ts(10, 0),
+            ts(11, 0),
+            ts(11, 0),
+            true,
+            false,
+            true,
+            uncommitted(10, TimeStamp::zero(), false),
+        );
+
+        // Path: the pessimistic primary key lock dose exist, and it's expired, the primary lock will
+        // be pessimistically rolled back but there will not be a rollback record.
+        must_success(
+            &engine,
+            k,
+            ts(10, 0),
+            ts(21, 0),
+            ts(21, 0),
+            true,
+            false,
+            true,
+            |s| s == PessimisticRollBack,
+        );
+        must_unlocked(&engine, k);
+        must_get_none(&engine, k, ts(22, 0));
+
+        // Path: the prewrite primary key lock dose exist, and it's not expired yet.
+        // Should return locked status.
+        must_prewrite_put_impl(
+            &engine,
+            k,
+            v,
+            k,
+            &None,
+            ts(30, 0),
+            false,
+            10,
+            TimeStamp::zero(),
+            1,
+            /* min_commit_ts */ TimeStamp::zero(),
+            /* max_commit_ts */ TimeStamp::zero(),
+        );
+        must_success(
+            &engine,
+            k,
+            ts(30, 0),
+            ts(31, 0),
+            ts(31, 0),
+            true,
+            false,
+            true,
+            uncommitted(10, TimeStamp::zero(), false),
+        );
+
+        // Path: the prewrite primary key expired and the solving key is a pessimistic lock,
+        // rollback record should be written and the transaction status is certain.
+        must_success(
+            &engine,
+            k,
+            ts(30, 0),
+            ts(41, 0),
+            ts(41, 0),
+            true,
+            false,
+            true,
+            |s| s == TtlExpire,
+        );
+        must_unlocked(&engine, k);
+        must_get_rollback_ts(&engine, k, ts(30, 0));
+    }
```

Comment:
Need one more case: `resolving_pessimistic_lock` is false and the primary key is a pessimistic lock. Rollback record should be written and the transaction status is certain.

Reply:
I think it's covered in L826.

Reply:
I think it's covered in L826